conn.go

```diff
@@ -325,10 +326,53 @@ func (c *Conn) Subprotocol() string {
 	return c.subprotocol
 }
 
-// Close closes the underlying network connection without sending or waiting
-// for a close message.
-func (c *Conn) Close() error {
-	return c.conn.Close()
+// Close sends close frame and waits for one in response
+// it expects two args. `closeCode int` and `closeMessage string` in order
+// it uses variadic args to maintain backwards compatibility
+func (c *Conn) Close(args ...interface{}) error {
+	closeCode := CloseNoStatusReceived
+	message := ""
+	ok := false
+	if len(args) == 2 {
+		closeCode, ok = args[0].(int)
+		if !ok {
+			closeCode = CloseNoStatusReceived
+		}
+		message, ok = args[1].(string)
+		if !ok {
+			message = ""
+		}
+	}
+	err := c.Shutdown(closeCode, message)
+	if err != nil {
+		return err
+	}
+	c.conn.Close()
+	return nil
+}
+
+// Shutdown sends a close frame and waits for one in response
+func (c *Conn) Shutdown(closeCode int, closeMessage string) error {
+	if !isValidCloseCode(closeCode) {
+		// we do not shutdown connection
+		return errors.New("invalid close code received")
+	}
+	if !utf8.ValidString(closeMessage) {
+		return  errors.New("invalid utf8 payload for shutdown message")
+	}
+	message := FormatCloseMessage(closeCode, closeMessage)
+	err := c.WriteControl(CloseMessage, message, time.Now().Add(writeWait))
+	if err != nil {
+		return err
+	}
+	timeStart := time.Now()
+	c.conn.SetReadDeadline(time.Now().Add(time.Minute))
+	for _, _, err := c.ReadMessage(); reflect.TypeOf(err) != reflect.TypeOf(&CloseError{}) ; {
```

Comment:
Use a type assertion instead of reflection.

Reply:
Thinking about this some more, the loop should break on any error.  Consider the case where the peer closed the connection.  This code will spin for 1000 iterations and then [panic](https://github.com/gorilla/websocket/blob/7c8e298727d149d7c329b4dec7e94e1932ac5c11/conn.go#L968).

Reply:
right. I missed the 1000 iterations case.